Feature/tgk 4 user curd

taekyun95 opened the pull request:
회원가입 기능 추가
세션 방식의 로그인, 로그아웃 기능 추가

taekyun95 committed:
#4 라이브러리 추가

spring-boot-starter-validation: 데이터 유효검 검사용
spring-boot-devtools: 개발편의성
mybatis-spring-boot-starter-test: Mybatis Test용
mapstruct: Object Mappering용

taekyun95 committed:
#4 설정추가 (DB, Mybatis)

Mybatis: 
  mapper-locations: 매퍼 위치 설정
  map-underscore-to-camel-case: 테이블 컬럼명을 스네이크 케이스를 커멀케이스로 매핑해주는 설정

taekyun95 committed:
#4 사용자 회원가입

taekyun95 committed:
#4 UserObjectMapper 오브젝트 매핑 설정 추가

id => 무시
createDate, updateDate => LocalDateTime.now()

taekyun95 committed:
#4 테스트 메소드명 한글로 변경

영어에서 한글로 변경

taekyun95 committed:
#4 기능: 세션 로그인 추가

taekyun95 committed:
#4 오타 수정

taekyun95 committed:
#4기능: 사용자 세션 로그아웃 추가

f-lab-jason commented on src/main/java/com/flab/commerce/mapper/UserMapper.java:
```diff
@@ -0,0 +1,15 @@
+
+    int insertUser(User newUser);
+
+
```
(nit) empty line은 늘 하나만 넣어주세요 :)

f-lab-jason commented on src/main/java/com/flab/commerce/mapper/UserMapper.java:
```diff
@@ -0,0 +1,15 @@
+@Mapper
+public interface UserMapper {
+
+    boolean existEmail(String email);
```
(nit) existence류의 boolean method 명은 보통 `emailExists()`가 제일 흔하게 쓰이는 convention입니다. (`isXXX()`나 `hasXXX()`에 비해 조금 갈리는 편이지만 그래도 이런쪽은 `xxxVerb()` 형태가 제일 흔히 보여요)

taekyun95 replied:
xxxVerb()형은 emailExists 같이 복수형으로 표현하나요?

bgpark82 replied:
보통의 메서드 명은 verbXXX 형태로 사용하는 것 같은데
이 부분은 어떻게 생각하시나요?

f-lab-jason replied:
복수형이 아니라 3인칭 동사로 표현합니다 :) (email이 단수이니)

f-lab-jason replied:
subjectVerb() 혹은 verbObject() 형태로 많이 씁니다. 메소드의 경우는 행위를 나타내니 verbObject()의 형태로 많이 쓰구요, 이부분은 subject의 존재 여부에 관한거니 subjectVerb()형태로 씁니다.

f-lab-jason commented on src/main/java/com/flab/commerce/user/UserController.java:
```diff
@@ -0,0 +1,54 @@
+@RequestMapping("/user")
+public class UserController {
+
+    public static final String USER_EMAIL = "USER_EMAIL";
```
공통으로 쓰이는 static String류는 한군데 모아주셔요. 보통 `Constants.java`같은 파일을 많이 씁니다.

bgpark82 replied:
그럼 하나의 클래스 안에서 아래처럼 나누는 방식을 사용하시나요?
```java
// user
public static final USER_CONSTANT = "USER_CONSTANT"

// owner
public static final OWNER_CONSTANT = "OWNER_CONSTANT"
```

f-lab-jason replied:
네 말씀하신것처럼 전반적으로 쓰이는 constants들은 안에서 영역만 나누어서 쓰는 경우가 많습니다.

f-lab-jason commented on src/main/java/com/flab/commerce/user/UserController.java:
```diff
@@ -0,0 +1,54 @@
+import lombok.RequiredArgsConstructor;
+
+import org.springframework.web.bind.WebDataBinder;
+import org.springframework.web.bind.annotation.*;
```
가급적이면 import시에 *는 피해주세요. import 라인이 늘어나더라도 사용하는 클래스만 import하는게 더 좋습니다.

bgpark82 replied:
혹시 이유를 알 수 있을까요?

개인적으로 추정하기는 다음과 같습니다.
1. 의존성을 한눈에 파악하기 어렵다
2. 의존성이 변경되었을 떄 merge 시 conflict 발생되서?

f-lab-jason replied:
제일 큰 이유는 클래스명이 겹치는 경우 때문입니다. *로 import할경우는 사용하지 않는 클래스도 같이 import되는데 다른 패키지에서 사용하는 클래스명과 겹치는 경우가 종종 발생해요.
물론 명시적인 의존성도 하나의 이유구요.

f-lab-jason commented on src/main/java/com/flab/commerce/user/UserController.java:
```diff
@@ -0,0 +1,54 @@
+        webDataBinder.addValidators(loginDtoValidator);
+    }
+
+
```
(nit) 2 empty lines

f-lab-jason commented on src/main/java/com/flab/commerce/user/UserController.java:
```diff
@@ -0,0 +1,54 @@
+    public void register(@Valid @RequestBody RegisterDto registerDto) {
+        User newUser = UserObjectMapper.INSTANCE.userToRegisterDto(registerDto);
+        userService.register(newUser);
+
```
(nit) empty line at the end

f-lab-jason commented on src/main/java/com/flab/commerce/user/User.java:
```diff
@@ -0,0 +1,21 @@
+    private String zipcode;
+    private String address;
+    private LocalDateTime createDate;
+    private LocalDateTime updateDate;
```
(nit) 날짜만 쓰실 예정이 아니면 `createDateTime` 같은 이름을 써주면 더 좋겠어요. `xxxDate`는 날짜만 나타낼때 쓰시기 바랍니다.

f-lab-jason commented on src/main/java/com/flab/commerce/user/dto/RegisterDto.java:
```diff
@@ -0,0 +1,31 @@
+    @NotBlank
+    private String password;
+    @NotBlank
+    private String confirmPassword;
```
서버사이드에서 `confirmPassword` 값이 필요한가요?

taekyun95 replied:
서버에서도 패스워드 검증을 하는게 안전하다라는 생각으로 추가했습니다.

f-lab-jason replied:
어떤 검증을 말씀하시는건가요? 보통 서버사이드에서 이런 검증은 필요하지 않습니다.
1. 프론트에서 password와 confirmPassword를 검증하고 요청을 보낼텐데 프론트에서 이미 검증한 confirmPassword를 보낼 필요가 있을까요?
2. 만약 현재 구조에서 password와 confirmPassword가 다 넘어왔는데 두 값이 다르다면 프론트를 거친 정상적인 요청일까요?

f-lab-jason commented on src/main/java/com/flab/commerce/user/dto/RegisterDto.java:
```diff
@@ -0,0 +1,31 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class RegisterDto {
```
모든 String property들에 대해 기본적인 길이 체크정도는 해주면 좋겠습니다. (다른 DTO들도 마찬가지구요)

f-lab-jason commented on src/main/java/com/flab/commerce/user/dto/RegisterDto.java:
```diff
@@ -0,0 +1,31 @@
+import javax.validation.constraints.Email;
+import javax.validation.constraints.NotBlank;
+
+
```
(nit) 2 empty lines

f-lab-jason commented on src/main/java/com/flab/commerce/user/validator/LoginDtoValidator.java:
```diff
@@ -0,0 +1,33 @@
+        if (user == null || !user.getPassword().equals(loginDto.getPassword())) result = true;
+
+        if (result) errors.reject("아이디 또는 비밀번호가 맞지 않습니다. 다시 확인해 주세요.");
+
```
(nit) empty line at the end

f-lab-jason commented on src/main/java/com/flab/commerce/user/validator/LoginDtoValidator.java:
```diff
@@ -0,0 +1,33 @@
+        boolean result = false;
+        if (user == null || !user.getPassword().equals(loginDto.getPassword())) result = true;
+
+        if (result) errors.reject("아이디 또는 비밀번호가 맞지 않습니다. 다시 확인해 주세요.");
```
1. if문은 가급적 늘 {}로 감싸서 new line으로 만들어주세요.
2. `errors.reject(...)`의 파라미터는 message가 아니라 errorCode이지 않나요?
3. 아래와 같은 형태가 더 나을 것 같은데 `result`를 따로 선언하신 이유가 있을까요?
```
if (user == null || !user.getPassword().equals(loginDto.getPassword())) {
    errors.reject("아이디 또는 비밀번호가 맞지 않습니다. 다시 확인해 주세요.");
}
```
4. `result`라는 이름이 굉장히 모호한 것 같네요. validation result라고 치면 validation이 잘된게 true인지 실패한게 true인지 알 수 없으니 이런 케이스는 보다 명확한 이름을 쓰면 좋겠습니다.

taekyun95 replied:
3. user null 검증과 password 검증을 통합하면서 남은 잔여 코드 입니다.
4. validation를 변수명으로 사용한다 명확히 표현되나요?

f-lab-jason replied:
3. 불필요한 로직은 남기지 않도록 해주세요 :)
4. 여기서 result의 용도가 무엇인가에 대해서 생각해보면 좋겠네요. isValid같은게 적당하겠네요.

f-lab-jason commented on src/main/java/com/flab/commerce/user/validator/RegisterDtoValidator.java:
```diff
@@ -0,0 +1,33 @@
+        if (userMapper.existEmail(registerDto.getEmail())) {
+            errors.rejectValue("email", "이미 존재하는 이메일입니다.");
+        }
+
```
(nit) empty line at the end

f-lab-jason commented on src/main/java/com/flab/commerce/user/UserController.java:
```diff
@@ -0,0 +1,54 @@
+    @PostMapping("/login")
+    public void login(@Valid @RequestBody LoginDto loginDto, HttpSession session) {
+        session.setAttribute(USER_EMAIL, loginDto.getEmail());
+    }
```
Validation result를 메소드 내부에서 처리해줘야할텐데 생략하신건가요?

taekyun95 replied:
Validation result 처리 부분이 비흡 했습니다.

f-lab-jason commented on src/main/java/com/flab/commerce/user/validator/RegisterDtoValidator.java:
```diff
@@ -0,0 +1,33 @@
+
+        if (!registerDto.getPassword().equals(registerDto.getConfirmPassword())) {
+            errors.rejectValue("confirmPassword", "비밀번호와 확인 비밀번호가 일치하지 않습니다.");
+        }
```
(nit) 이 파트는 보통 frontend가 처리하니 backend에서는 굳이 안하셔도 될 것 같아요.

f-lab-jason commented on src/main/resources/mappers/UserMapper.xml:
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<!DOCTYPE mapper PUBLIC "-//mybatis.org//DTD Mapper 3.0//EN" "http://mybatis.org/dtd/mybatis-3-mapper.dtd" >
+<mapper namespace="com.flab.commerce.mapper.UserMapper">
```
(nit) SQL예약어(SELECT, INTO, FROM, WHERE 등등)은 주로 대문자로 써줍니다.

f-lab-jason commented on src/main/resources/mappers/UserMapper.xml:
```diff
@@ -0,0 +1,18 @@
+    <select id="findByEmail" resultType="com.flab.commerce.user.User">
+        select *
+        from user
+        where email = #{email};
```
(nit) SQL뒤에 ;를 붙히는걸 선호하시면 모두 붙혀주시고 아니면 모두 제거해주세요 :)

bgpark82 replied:
말씀대로 통일하는게 좋을 것 같네요

f-lab-jason commented on src/test/java/com/flab/commerce/user/UserControllerTest.java:
```diff
@@ -0,0 +1,179 @@
+    @Autowired
+    WebApplicationContext context;
+
+
```
(nit) 2 empty lines

f-lab-jason commented on src/test/java/com/flab/commerce/user/UserControllerTest.java:
```diff
@@ -0,0 +1,179 @@
+                        .content(json))
+                .andDo(print())
+                .andExpect(status().isOk());
+
```
(nit) empty line at the end

f-lab-jason commented on src/test/java/com/flab/commerce/user/UserControllerTest.java:
```diff
@@ -0,0 +1,179 @@
+    @Test
+    void 회원가입_실패_이메일_중복() throws Exception {
+
+        when(userMapper.existEmail("teset@gmail.com")).thenReturn(true);
```
(nit) 'teset'@gmail.com은 의도하신건가요?

taekyun95 replied:
오타입니다

f-lab-jason commented on src/test/java/com/flab/commerce/user/UserControllerTest.java:
```diff
@@ -0,0 +1,179 @@
+                .andExpect(status().isBadRequest());
+    }
+
+
```
(nit) 2 empty lines

f-lab-jason commented on src/test/java/com/flab/commerce/user/UserControllerTest.java:
```diff
@@ -0,0 +1,179 @@
+
+    @Test
+    void 로그아웃_성공() throws Exception {
+
```
(nit) empty line at the start

f-lab-jason commented on src/test/java/com/flab/commerce/user/UserMapperTest.java:
```diff
@@ -0,0 +1,62 @@
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.dao.DuplicateKeyException;
+
+
```
(nit) 2 empty lines

f-lab-jason commented on src/test/java/com/flab/commerce/user/UserMapperTest.java:
```diff
@@ -0,0 +1,62 @@
+                .phone("010-1234-5678")
+                .password("1234")
+                .build();
+
```
(nit) empty line at the end

f-lab-jason commented on src/test/java/com/flab/commerce/user/UserServiceTest.java:
```diff
@@ -0,0 +1,42 @@
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.BDDMockito.given;
+
+
```
(nit) 2 empty lines

f-lab-jason commented on src/main/java/com/flab/commerce/mapper/UserMapper.java:
```diff
@@ -0,0 +1,15 @@
+
+@Mapper
+public interface UserMapper {
+
```
(nit) 불필요한 empty line

taekyun95 replied:
(nit)에 의미가 무엇인지 알 수 있을까요??

f-lab-jason replied:
nit picking이라고 사소한 수정사항이라는 뜻입니다. 실제 로직상에서는 크게 영향이 없고 주로 convention관련된 내용이나 개인선호에따라 갈리는 부분을 얘기해요.

f-lab-jason commented on build.gradle:
```diff
@@ -21,11 +21,17 @@ repositories {
 dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-web'
     implementation 'org.mybatis.spring.boot:mybatis-spring-boot-starter:2.2.2'
+    implementation 'org.springframework.boot:spring-boot-starter-validation'
     compileOnly 'org.projectlombok:lombok'
-    runtimeOnly 'com.h2database:h2'
+
     runtimeOnly 'mysql:mysql-connector-java'
+    developmentOnly 'org.springframework.boot:spring-boot-devtools'
     annotationProcessor 'org.projectlombok:lombok'
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
+    testImplementation  "org.mybatis.spring.boot:mybatis-spring-boot-starter-test:2.2.2"
+    implementation 'org.mapstruct:mapstruct:1.5.1.Final'
+    annotationProcessor 'org.mapstruct:mapstruct-processor:1.5.1.Final'
+
 }
```
(nit) 이런 dependency류를 선언할땐 같은 namespace끼리(implementation, testImplementation등) 모아주시고 뒤에 적는것도 가급적이면 lexicographic으로 선언해주시면 좋습니다. formatter가 있을텐데 그걸 쓰는것도 좋구요. 그룹사이엔 한 줄씩 공백을 주시기 바랍니다.

f-lab-jason reviewed (commented):
1. (nit) (commit msg) 보통 CRUD가 common term 입니다 :)
2. 시간류에 `LocalDateTime`을 쓰신 이유가 있을까요? System clock을 따라가는거라 하드웨어따라 다른 시간을 받을 가능성이 있습니다.
3. 병길님과 상의하여 프로젝트 전반에 쓰일 code formatter를 하나 선택하신후에 일괄 적용하면 좋겠어요 :)
4. 리뷰 요청시엔 병길님도 같이 부탁드려요!

taekyun95 committed:
코드 리뷰 반영

1. * import 풀어 사용하기
2. static String을 한 곳에서 모이기
3. 구분의 의미 없는 공백 제거
4. DTO String 속성 길이 제한
5. DateTime ​사용하는 변수에 이름을 정확히 표현
6. gradle 의존성 namespace를 그룹핑 + 사전 순 정렬

taekyun95 committed:
#4 코드 리뷰 반영

1. * import 풀어 사용하기
2. static String을 한 곳에서 모이기
3. 구분의 의미 없는 공백 제거
4. DTO String 속성 길이 제한
5. DateTime ​사용하는 변수에 이름을 정확히 표현
6. gradle 의존성 namespace를 그룹핑 + 사전 순 정렬

taekyun95 committed:
Merge remote-tracking branch 'origin/feature/tgk_4_user_curd' into feature/tgk_4_user_curd

taekyun95 committed:
#4 JSON 방식 스프링 시큐리티 적용

기타
RegisterDto: @Setter 제거
UserController: 로그인, 로그아웃 제거
UserObjectMapper: password 인코딩 매핑 추가
LoginDtoValidator: 제거
JUnit Assertion => AssertJ

taekyun95 committed:
세션 저장소 변경: 로컬 저장소에서 레디스로

taekyun95 committed:
변경 LocalDateTime에서 ZonedDateTime으로

taekyun95 committed:
변경 클래스명: UserContext에서 RestUserDetails로

bgpark82 commented on src/main/resources/mappers/UserMapper.xml:
```diff
@@ -0,0 +1,12 @@
+    </insert>
+
+    <select id="existEmail" resultType="boolean">
+        select exists(select email from user where email = #{email})
```
select exists라는 문법은 mysql에 의존적인 문법이 아닐까요?

bgpark82 commented on src/main/java/com/flab/commerce/user/UserController.java:
```diff
@@ -0,0 +1,32 @@
+public class UserController {
+
+    private final UserService userService;
+    private final RegisterDtoValidator registerDtoValidator;
```
두가지 질문이 있습니다.

1. Validation 로직은 Presentation 레이어 보다는 Service 레이어의 책임이 더 맞지 않을까?
2. 기존에 Validation 로직을 Service 레이어에서 작성했었는데 Validation 클래스로 따로 빼서 관리하면 의존성이 나눠지는 게 아닌가 궁금합니다

bgpark82 commented on src/main/java/com/flab/commerce/user/UserController.java:
```diff
@@ -0,0 +1,32 @@
+
+    @PostMapping("/user/register")
+    public void register(@Valid @RequestBody RegisterDto registerDto) {
+        User newUser = UserObjectMapper.INSTANCE.userToRegisterDto(registerDto);
```
혹시 Mapper를 더 잘 사용하는 방법이 있을지 궁금합니다.

지금은 메소드 내부에 Mapper가 하드 코딩 된 상태인데
이렇게 되면 Mapper 로직이 변경되면 Mapper 변경이 필요하는 곳 마다 코드를 수정해야 될 것 같아서요

taekyun95 committed:
#4 로그인, 로그아웃, 회원가입 URL 규칙 적용

taekyun95 committed:
#4 시큐리티 클래스 명칭 수정

RestXXX => GeneralXXX

taekyun95 committed:
#4 Validation 테스트 추가

taekyun95 committed:
#4 AuthenticationProcessingFilter에 validation추가

taekyun95 committed:
공백 정리
